src/Tools/Source/RunTests/TestHistoryManager.cs

```diff
@@ -0,0 +1,213 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.Build.WebApi;
+using Microsoft.TeamFoundation.TestManagement.WebApi;
+using Microsoft.VisualStudio.Services.TestResults.WebApi;
+using Microsoft.VisualStudio.Services.WebApi;
+
+namespace RunTests;
+internal class TestHistoryManager
+{
+    /// <summary>
+    /// Azure devops limits the number of tests returned per request to 10000.
+    /// </summary>
+    private const int MaxTestsReturnedPerRequest = 10000;
+
+    /// <summary>
+    /// The pipeline id for roslyn-ci, see https://dev.azure.com/dnceng/public/_build?definitionId=15
+    /// </summary>
+    private const int RoslynCiBuildDefinitionId = 15;
+
+    /// <summary>
+    /// The Azure devops project that the build pipeline is located in.
+    /// </summary>
+    private static readonly Uri s_projectUri = new(@"https://dev.azure.com/dnceng");
+
+    /// <summary>
+    /// Looks up the last passing test run for the current build and stage to estimate execution times for each test.
+    /// </summary>
+    public static async Task<ImmutableDictionary<string, TimeSpan>> GetTestHistoryAsync(CancellationToken cancellationToken)
+    {
+        // Gets environment variables set by our test yaml templates.
+        // The access token is required to lookup test histories.
+        // We use the target branch of the current build to lookup the last successful build for the same branch.
+        //
+        // The phase name is used to filter the tests on the last passing build to only those that apply to the currently running phase.
+        //   Note here that 'phaseName' corresponds to the 'jobName' defined in our pipeline yaml file and the job name env var is not correct.
+        //   See https://developercommunity.visualstudio.com/t/systemjobname-seems-to-be-incorrectly-assigned-and/1209736
+        if (!TryGetEnvironmentVariable("SYSTEM_ACCESSTOKEN", out var accessToken)
+            || !TryGetEnvironmentVariable("SYSTEM_PHASENAME", out var phaseName))
+        {
+            Console.WriteLine("Missing required environment variables - skipping test history lookup");
```

Comment:
Several uses of  `Console` here look like they should be `ConsoleUtil` so we get command line + log file.